### PR TITLE
Resolves #19790, Provides Support for IAM Credentials

### LINKED ISSUE #19790: Enhance S3 to support EC2 instance roles
**Is your feature request related to a problem? Please describe.**
At the moment, when you configure NC to support S3 as primary object storage, you need to specify the access key ID and secret access key. This, in turn, requires the creation of a user on IAM. The drawback to this is that the keys get stale over time which becomes a security vulnerability.

**Describe the solution you'd like**
If NC is running on an AWS EC2 instance, it should be possible to use the instance role to gain the appropriate credentials. The code could try to use this mechanism automatically if `config.php` is missing the keys.

https://docs.aws.amazon.com/aws-sdk-php/v2/guide/credentials.html#instance-profile-credentials

**Describe alternatives you've considered**
Creating a user and then manually rotating the keys regularly but this requires the config file to get updated and NC to be restarted.

### PR DESCRIPTION
**Pull request by cuppett**
Includes support for either leveraging environment variables passed to the PHP runtime or IAM instance profile present on the host being used. The default and first choice is still the parameter file as documented.

See also: [Developer Guide -> Chaining Providers](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials_provider.html#chaining-providers)

### CODE REVIEW

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php [hunk 1]
```diff
@@ -156,4 +169,21 @@ public static function legacySignatureProvider($version, $service, $region) {
 				return null;
 		}
 	}
+
+	// This function CREATES a credential provider based on user parameter file
+	protected function paramCredentialProvider() {
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/S3ConnectionTrait.php:174]:
```suggestion
	protected function paramCredentialProvider(): callable
```

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php [hunk 2]
```diff
@@ -156,4 +169,21 @@ public static function legacySignatureProvider($version, $service, $region) {
 				return null;
 		}
 	}
+
+	// This function CREATES a credential provider based on user parameter file
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/S3ConnectionTrait.php:173]:
```suggestion
    /**
     * This function creates a credential provider based on user parameter file
     */
```

**Review by kesselb — COMMENTED** [review]
Thanks :+1:

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php (hunk 1 above)
Comment (cuppett) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Done.

File: lib/private/Files/ObjectStore/S3ConnectionTrait.php (hunk 2 above)
Comment (cuppett) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Done.

Comment (kesselb) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Thx. Comment for method in php is usually

/**
  *
  */

Comment (cuppett) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Oof. Now done.

Comment (kesselb) [reply lib/private/Files/ObjectStore/S3ConnectionTrait.php]: Thanks :+1:

**Comment by cuppett** [comment]
Anything else needed from me here? Is there a way for me to re-launch/attempt the failing CI build (or is it a common failure)?

**Comment by kesselb** [comment]
> Anything else needed from me here

I don't think so.

> for me to re-launch/attempt the failing CI build 

Rebase would be nice.
 
> is it a common failure

Yes.

**Comment by cuppett** [comment]
> > Anything else needed from me here
> 
> I don't think so.
> 
> > for me to re-launch/attempt the failing CI build
> 
> Rebase would be nice.
> 
> > is it a common failure
> 
> Yes.

Thanks! Rebase done.

**Comment by MorrisJobke** [comment]
Rebased because we had some broken test on master today. Sorry for that.

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by MorrisJobke** [comment]
Let me rebase again and @rullzer will review later today.

**Comment by kesselb** [comment]
Psalm warnings looks like a false positive to me. Should work if we import the function `use function GuzzleHttp\Promise\promise_for;` and use it like

```
				return promise_for(
					new Credentials($key, $secret)
				);
```

**Comment by cuppett** [comment]
> Psalm warnings looks like a false positive to me. Should work if we import the function `use function GuzzleHttp\Promise\promise_for;` and use it like
> 
> ```
> 				return promise_for(
> 					new Credentials($key, $secret)
> 				);
> ```

I have that cleanup... squash and force push or add commit?

**Review by rullzer — APPROVED** [review]
Still works when running with minio!
Good stuff!

**Comment by MorrisJobke** [comment]
> I have that cleanup... squash and force push or add commit?

Both are fine ... @cuppett mind to send another PR ... it seems @rullzer merged too early :see

**Comment by MorrisJobke** [comment]
https://github.com/nextcloud/server/pull/22345

**Pull request merged**